### PR TITLE
⭐ feat(rules/number-literal-format): add fix

### LINKED ISSUE #3970: number-literal-format autofix

### PR DESCRIPTION
**Pull request by ColCh**
#### PR checklist

- [x] Addresses an existing issue: fixes #3970
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [ ] Documentation update

#### Overview of change:

Added fix for __number-format__ rule

#### Is there anything you'd like reviewers to focus on?

tricky test cases, corner cases

#### CHANGELOG.md entry:

[new-fixer] `number-literal-format` now includes auto fix

### CODE REVIEW

File: test/rules/number-literal-format/test.ts.lint [hunk 1]
```diff
@@ -3,38 +3,41 @@
...
 
-.50;
-~~~ [trailing-0]
-~~~ [leading-decimal]
+-.2;
+ ~~ [leading-decimal]
 
-1e01;
-  ~~ [leading-0]
-1E01;
```
Comment (JoshuaKGoldberg) [inline test/rules/number-literal-format/test.ts.lint]: Is there a reason you're removing valid syntax tests such as `1e01;` and `1E01;`? In the case of folks not using Prettier, we'd still want to validate these cases working.

Comment (ColCh) [reply test/rules/number-literal-format/test.ts.lint]: Oops. Returned that tests back

**Comment by ColCh** [comment]
@JoshuaKGoldberg Please take a look again :)

File: src/rules/numberLiteralFormatRule.ts [hunk 2]
```diff
@@ -82,35 +90,60 @@ function walk(ctx: Lint.WalkContext<void>): void {
...
+                        Lint.Replacement.deleteFromTo(start, start + /^0+/.exec(text)![0].length),
+                    );
                     return;
             }
         }
 
-        const [num, exp] = text.split(/e/i);
+        const [num, exp = ""] = text.split(/e/i);
+        const [integer, float = ""] = num.split(".");
+        const match = /(\.)([1-9]*)(0+)/.exec(num);
```
[the remark this answers is not in the document]
Comment (ColCh) [reply src/rules/numberLiteralFormatRule.ts]: This one is tricky 😁 I can't find cohesive name too

changed to `matchedNumeric`

File: src/rules/numberLiteralFormatRule.ts [hunk 3]
```diff
@@ -82,35 +90,60 @@ function walk(ctx: Lint.WalkContext<void>): void {
...
+        const [dot = "", numbers = "", zeroes = ""] = Array.isArray(match) ? match.slice(1) : [];
+
         if (exp !== undefined && (exp.startsWith("-0") || exp.startsWith("0"))) {
```
[the remark this answers is not in the document]
Comment (ColCh) [reply src/rules/numberLiteralFormatRule.ts]: Oh, yes. Didn't noticed that. Fixed!

File: src/rules/numberLiteralFormatRule.ts [hunk 4]
```diff
@@ -82,35 +90,60 @@ function walk(ctx: Lint.WalkContext<void>): void {
...
             return;
-        }
-
-        if (num.startsWith(".")) {
-            fail(Rule.FAILURE_STRING_LEADING_DECIMAL);
-        }
-
-        if (num.endsWith(".")) {
-            fail(Rule.FAILURE_STRING_TRAILING_DECIMAL);
+        } else if (num.startsWith(".")) {
```
[the remark this answers is not in the document]
Comment (ColCh) [reply src/rules/numberLiteralFormatRule.ts]: Just wanted to shrink `if`'s a bit, but wrote a nit instead 🤔Fixed, thanks!

**Comment by ColCh** [comment]
Applied proposed changes. Please revisit

**Comment by ColCh** [comment]
@JoshuaKGoldberg Thank you for review! 👍

**Comment by rgant** [comment]
I just had this fixer fail on a number with leading and trailing zeros:

```
-            rate: 0.022000,
+            rate: 22000,
```

**Comment by JoshuaKGoldberg** [comment]
@rgant oh! It'd be swell if you could [file a bug report](https://github.com/palantir/tslint/issues/new?template=bug_report.md) for that!

**Comment by rgant** [comment]
I will, but because I'm using Angular I'm behind in tslint versions so it doesn't seem great to file a bug that would hopefully just be responded to with "try to reproduce the latest version". The playground doesn't appear to have the ability to run the fixer so I need to setup the env. Sorry for creating noise here.
